### PR TITLE
Add Solidity-based console logging in test environment

### PR DESCRIPTION
**Pull request by eggplantzzz**
This PR is meant to be a continuation of #2862 

Original description from @adrianmcli:

> We should no longer hold this feature back from our users.
> 
> This PR puts `Console.log(variable)` and `Console.log("myVarLabel", variable)` behind a feature flag into Truffle so people can use it without downloading a different version of Truffle.
> 
> Simply:
> 
> 1. `import "truffle/Console.sol";`
> 2. `Console.log(variable)`
> 3. Run your Truffle tests and see the result

In the future we may make `Console` available in other environments but currently this will only be available in the test environment.

### CODE REVIEW

File: packages/core/lib/testing/TestSource.js
Comment (gnidan) [inline packages/core/lib/testing/TestSource.js:112]: Did I accidentally add this here when you removed it @eggplantzzz?

File: packages/migrate/Migration.js
Comment (gnidan) [inline packages/migrate/Migration.js:221]: Is it okay that we just let this fail?

File: packages/resolver/lib/sources/truffle.ts
Comment (gnidan) [inline packages/resolver/lib/sources/truffle.ts:16]: I don't like how this is reaching into @truffle/core... especially considering that we might want to do the same here with Assert+DeployedAddresses. I think those contracts need a new home. Any ideas? I was thinking we could have @truffle/solidity-logging for Console

File: packages/core/lib/testing/TestSource.js
Comment (eggplantzzz) [reply packages/core/lib/testing/TestSource.js:112]: No, I think this is where this was.

Comment (gnidan) [reply packages/core/lib/testing/TestSource.js:112]: Hm, then seems it's duplicated by your Truffle resolver source, no? Is there a way to remove this duplication?

Comment (eggplantzzz) [reply packages/core/lib/testing/TestSource.js:112]: Ok let me check on it again to see which is correct. Maybe I'm mistaken

File: packages/migrate/Migration.js
Comment (eggplantzzz) [reply packages/migrate/Migration.js:221]: I suppose we probably should let this fail, or at the very least warn them? It seems better since they have the logging turned on and it would fail.

File: packages/core/lib/testing/TestSource.js
Comment (eggplantzzz) [reply packages/core/lib/testing/TestSource.js:112]: It is duplicated but I'm not sure it is worth it to deal with such a small amount of code. Anything come to mind?

File: packages/resolver/lib/sources/truffle.ts
Comment (gnidan) [reply packages/resolver/lib/sources/truffle.ts:16]: We could, it just feels gross to have them all there. Maybe the new package could be `@truffle/solidity-libs`, for all three?

File: packages/migrate/Migration.js
Comment (gnidan) [reply packages/migrate/Migration.js:221]: It might be the right behavior to abort like this. I'm just wondering if there's a way we can detect certain allowed failure conditions, rather than catch-all aborting every kind of exception with no notice. Like, surely, in some situations where things break, we should alert the user, no?

File: packages/core/lib/testing/TestSource.js
Comment (gnidan) [reply packages/core/lib/testing/TestSource.js:112]: It just makes me wonder... is your Truffle ResolverSource even getting used?

File: packages/resolver/lib/sources/truffle.ts
Comment (eggplantzzz) [reply packages/resolver/lib/sources/truffle.ts:16]: It seems to me that all these would be at home in `resolver`. They are basically libraries that the `truffle` source resolves. That feels pretty appropriate to me. What do you think?

File: packages/migrate/Migration.js
Comment (eggplantzzz) [reply packages/migrate/Migration.js:221]: So really this code should never throw unless there is an error on our side when a user is using Truffle, right? I suppose if they are using the migration library stand-alone and don't configure something correctly it might fail as well. I might be in favor of just printing a warning to the screen that "you are using the solidity logging flag but we cannot resolve the library" or something like that.

File: packages/core/lib/testing/TestSource.js
Comment (eggplantzzz) [reply packages/core/lib/testing/TestSource.js:112]: Currently, it looks to me like `TestSource` will always resolve this logging library and it will never get it from `resolver`.

Comment (eggplantzzz) [reply packages/core/lib/testing/TestSource.js:112]: So the conclusion here was to remove `TestSource` altogether since that is now handled by the `truffle` source.

File: packages/resolver/lib/sources/truffle.ts
Comment (eggplantzzz) [reply packages/resolver/lib/sources/truffle.ts:16]: The decision was made to move all these templates into the `resolver` package.

**Review by fainashalts — APPROVED** [review]
A couple of comments but overall LGTM!

**Comment by eggplantzzz** [comment]
Eventually we'll want to re-organize some of this stuff to agree with the organization of https://github.com/trufflesuite/truffle/pull/3395. Perhaps we should wait until that is merged.

**Review by gnidan — CHANGES_REQUESTED** [review]
Marking as request-changes as per comment above re: #3395

File: packages/contract/lib/utils/index.js
Comment (eggplantzzz) [reply packages/contract/lib/utils/index.js:114]: Hmm, I can't really think of an elegant way of doing this other than creating an awkward function because both of these are in closures. In addition, you'd have to pass the index as well which would be ugly. Any ideas?

File: packages/migrate/Migration.js
Comment (eggplantzzz) [reply packages/migrate/Migration.js]: So this function got deleted and it now gets deployed and linked like all of the other testing libraries in  `packages/core/lib/testing/SolidityTest.js`.

File: packages/resolver/solidity/Console.sol
Comment (gnidan) [inline packages/resolver/solidity/Console.sol:1]: Do we need to update this to support `^0.7.0` now?

**Review by gnidan — COMMENTED** [review]
Just went through all this again. Looks good!

Looks like the console stuff doesn't support 0.7.0; we probably want to fix that? But apart from that, this looks good to go!

File: packages/resolver/solidity/Console.sol
Comment (eggplantzzz) [reply packages/resolver/solidity/Console.sol:1]: Yeah I suppose this should have a `<0.8.0`, right?

Comment (gnidan) [reply packages/resolver/solidity/Console.sol:1]: yep and it might have some other changes required

Comment (eggplantzzz) [reply packages/resolver/solidity/Console.sol]: updated  the pragma

File: packages/contract/lib/utils/index.js
Comment (eggplantzzz) [reply packages/contract/lib/utils/index.js:131]: Yeah I'm not sure, we should have a discussion at some point about how we want to deal with all the logging on a large scale in Truffle.

**Review by gnidan — APPROVED** [review]
I think we can call this one good now!

**Pull request merged**
